frmts/tiledb/tiledbdataset.cpp

```diff
@@ -298,278 +351,114 @@ TileDBRasterBand::TileDBRasterBand(TileDBDataset *poDSIn, int nBandIn,
+                new tiledb::Query(*poGDS->m_roCtx, *poGDS->m_roArray));
         }
-        catch (const std::exception &e)
+        else
         {
-            CPLError(CE_Failure, CPLE_AppDefined,
-                     "TileDBRasterBand::IWriteBlock(): set_subarray() failed "
-                     "with %s",
-                     e.what());
-            return CE_Failure;
+            poQuery.reset(new tiledb::Query(*poGDS->m_ctx, *poGDS->m_array));
         }
-    }
-    else
-    {
-        // global order requires ordered blocks (see FlushCache(bool
-        // bAtClosing))
-        if ((nCurrBlockX != nBlockXOff) || (nCurrBlockY != nBlockYOff))
+
+        if (poGDS->m_array->schema().domain().ndim() == 3)
         {
-            CPLError(CE_Failure, CPLE_AppDefined,
-                     "Non-sequential global write to TileDB.\n");
-            return CE_Failure;
+            poQuery->set_subarray(oaSubarray);
         }
         else
         {
-            if (++nCurrBlockX == poGDS->nBlocksX)
-            {
-                nCurrBlockX = 0;
-                nCurrBlockY++;
-            }
+            poQuery->set_subarray(std::vector<uint64_t>(oaSubarray.cbegin() + 2,
+                                                        oaSubarray.cend()));
         }
-    }
 
-    std::vector<std::unique_ptr<void, decltype(&VSIFree)>> aBlocks;
+        SetBuffer(poQuery.get(), eDataType, osAttrName, pData, nXSize * nYSize);
 
-    if (poGDS->lpoAttributeDS.size() > 0)
-    {
-        for (auto const &poAttrDS : poGDS->lpoAttributeDS)
+        // write additional co-registered values
+        std::vector<std::unique_ptr<void, decltype(&VSIFree)>> aBlocks;
+
+        if (poGDS->lpoAttributeDS.size() > 0)
         {
-            GDALRasterBand *poAttrBand = poAttrDS->GetRasterBand(nBandIdx);
-            GDALDataType eAttrType = poAttrBand->GetRasterDataType();
-            int nBytes = GDALGetDataTypeSizeBytes(eAttrType);
-            int nValues = nBlockXSize * nBlockYSize;
-            void *pAttrBlock = VSIMalloc(nBytes * nValues);
-            aBlocks.emplace_back(pAttrBlock, &VSIFree);
-
-            if (pAttrBlock == nullptr)
+            for (auto const &poAttrDS : poGDS->lpoAttributeDS)
             {
-                CPLError(CE_Failure, CPLE_OutOfMemory,
-                         "Cannot allocate attribute buffer");
-                return CE_Failure;
-            }
-
-            int nXSize = nBlockXSize;
-            int nYSize = nBlockYSize;
-            if (nBlockXOff + nXSize > nRasterXSize)
-                nXSize = nRasterXSize - nBlockXOff;
-            if (nBlockYOff + nYSize > nRasterYSize)
-                nYSize = nRasterYSize - nBlockYOff;
+                GDALRasterBand *poAttrBand = poAttrDS->GetRasterBand(nBandIdx);
+                GDALDataType eAttrType = poAttrBand->GetRasterDataType();
+                int nBytes = GDALGetDataTypeSizeBytes(eAttrType);
+                int nValues = nBufXSize * nBufYSize;
```

Comment:
To avoid potential overflows:
```suggestion
                size_t nValues = static_cast<size_t>(nBufXSize) * nBufYSize;
```

Reply:
change made

Reply:
change made